### ISSUE TITLE
Move testing requirements to the dialects

### ISSUE THREAD

**Issue by CaselIT** [sqlalchemy/sqlalchemy#5174]
The testing framework used in sqlalchemy requires that some tests are only run on particular database backend, since not all features are supported by every db.

Currently this functionality is obtained using the file `test/requirement.py` and `lib/sqlalchemy/testing/requirements.py`

To improve the experience for client libraries and external dialects, this functionality should be moved in a `requirement.py` file in each dialect folder, so that these requirements are shipped with sqlalchemy and can be leveraged by client libraries in its tests. Alembic, for example, has a similar need to check for requirements in the tests and needs to re-implement most of the same checks that are in sqlalchemy requirements.

This refactoring is similar to the provision functionality, that was moved to the dialects in https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1663

**Comment by zzzeek** [comment]
don't have a good keyword here.   the 1.5 milestone can be 1.4 if the work got done but there is enough going on already.

**Comment by gordthompson** [comment]
I did a little digging around to see what this can of worms might entail. Here's what I found. As shorthand I'll use 

- "test_req" for "test/requirements.py",
- "testing_req" for "lib/sqlalchemy/testing/requirements.py", and
- "dialect_req" for "lib/sqlalchemy/dialects/oracle/requirements.py" (for example).

*(Aside: "requirements.py" is a rather ambiguous file name, especially when located in a subfolder that is not part of a "test[ing]" tree. Perhaps "test_requirements.py" might be a better name?)*

By sheer coincidence, both files contain 141 @property definitions. 92 of them are common to both files, so both test_req and testing_req contain 49 properties that are specific to that file.

[common.txt](https://github.com/sqlalchemy/sqlalchemy/files/4278538/common.txt)
[test_only.txt](https://github.com/sqlalchemy/sqlalchemy/files/4278542/test_only.txt)
[testing_only.txt](https://github.com/sqlalchemy/sqlalchemy/files/4278543/testing_only.txt)

### Common requirements

I'm guessing that testing_req will become the one source of "reasonable defaults" for all requirements (ref: README.dialects.rst). Therefore any default-y requirements in test_req should be moved into testing_req with dialect-specific values moved into dialect_req. For example, this will stay the same ...

[testing_req]
```python
    @property
    def ad_hoc_engines(self):
        """Test environment must allow ad-hoc engine/connection creation.

        DBs that scale poorly for many connections, even when closed, i.e.
        Oracle, may use the "--low-connections" option which flags this
        requirement as not present.

        """
        return exclusions.skip_if(
            lambda config: config.options.low_connections
        )
```

... and this will be removed from test_req ...

```python
    @property
    def ad_hoc_engines(self):
        return exclusions.skip_if(
            ["oracle"],
            "works, but Oracle just gets tired with "
            "this much connection activity",
        )
```

... to be replaced by this in Oracle's dialect_req:

```python
    @property
    def ad_hoc_engines(self):
        # works, but Oracle just gets tired with this much connection activity
        return exclusions.closed()
```

### Dialect-specific requirements

Presumably any dialect-specific requirements like

```python
    @property
    def no_mssql_freetds(self):
        return self.mssql_freetds.not_()
```

will have no default requirement in testing_req, it will just move to dialect_req.

Does it sound like I'm on the right track?

**Comment by zzzeek** [comment]
yeah sounds like you're doing great.

I think there just needs to be some extra thing that in testing_req can look at "config.db.dialect" and then pull in the dialect_req from that point and call the right method.

**Comment by CaselIT** [comment]
>  testing_req can look at "config.db.dialect" and then pull in the dialect_req

Should this only happen while testing, or should dialect always have a way to get it's requirements?
I mean is this code supposed to work outside tests?:
```py
from sqlalchemy.dialects.postgres import dialect
pg = dialect()
pg.requirements
```

**Comment by zzzeek** [comment]
i think at the moment we probably dont need to go there yet.    I can see asking the question, hey is this whole requirements thing actually just a runtime registry for anyone to use, and I think the answer is, not at the moment.  the functions in the  Requirements class are extremely specific to test suites and the specific kinds of tests that are used here.   To the degree that a dialect can tell you what a database backend can and can't do, that's still just the flags that are on the Dialect itself.  the requirements thing, which at this point we can likely agree is not named very well (it pre-dates pip and requirements files), interprets the Dialect and lots of other things in order to determine rules for running tests.

**Comment by CaselIT** [comment]
Yes I agree that it is probably not very useful to have it at runtime. The question was because you suggested looking for it in the dialect.
Regarding the name I came up with `prerequisites` or `testing_prerequisites` just now (but I'm usually pretty bad at naming things)
